File: algorithms/elife.js

// plan outlines what a world willlook like
// # is terrain, o is a critter
var plan = ["############################",
			"#                          #",
			"#                     o    #",
			"#                          #",
			"#         ##       ##      #",
			"#         ##       ##      #",
			"#                          #",
			"#     o                    #",
			"#                          #",																					
			"############################"]

// these functions help represent space
function Vector(x, y){
	this.x = x;
	this.y = y;
}

// for adding vectors
Vector.prototype.plus = function(other){
	return new Vector(this.x + other.x, this.y + other.y);
};

// to represent space using a single array(witha a base^width map)
// to define an actual array using grid, 
function Grid(width, height){
	this.space = new Array(width * height);
	this.width = width;
	this.height = height;
}
Grid.prototype.isInside = function(Vector){
	return  vector.x >= 0 && vector.x < this.width &&
			vector.y >= 0 && vector.y < this.height;
};
Grid.prototype.get = function(vector){
	return this.space[vector.x + this.width*vector.y];
};
Grid.prototype.set = function(vector, value){
	this.space[vector.x + this.width*vector.y] = value;
};

// some critter work

// let them see their surroundings, ye?
var directions = {
	"n": 	new Vector( 0,-1),
	"ne": 	new Vector( 1,-1),
	"e": 	new Vector( 1, 0),
	"se": 	new Vector( 1, 1),
	"s": 	new Vector( 0, 1),
	"sw": 	new Vector(-1, 1),
	"w": 	new Vector(-1, 0),
	"nw": 	new Vector(-1,-1),
};

// the following helps make a random critter that plods along 
// and then bounces off in a random direction

// finds a place to start lookiing
function randomElement(array){
	return array[Math.floor(Math.random()*array.length)];
}
// lets it understand the directions
var directionNames = "n ne e se s sw w nw".split(" ");

// bouncing critter onlyknows what direction is looking
function BouncingCritter(){
	this.direction = randomElement(directionNames);
}

// action it has is to move or bounce
BouncingCritter.prototype.act = function(view){
	if (view.look(this.direction) != " "){
		this.direction = view.find(" ") || "s";
	}
	return {type: "move", direction: this.direction};
};


// some world work
// this lets us read the plan, or map as it is
function elementFromChar(legend, ch){
	if (ch == " "){return null;}
	var element = new legend[ch]();
	element.originChar = ch;
	return element;
}

function charFromElement(element){
	if (element == null){ return " ";}
	else { return element.originChar;}
}

// creates our world array, using a map and its 
// array elements to determine width and height
function World(map, legend){
	var grid = new Grid(map[0].length, map.length);
	this.grid = grid;
	this.legend = legend;

	map.forEach(function(line, y){
		for (var x = 0; x < line.length; x++){
			grid.set(new Vector(x,y),
				elementFromChar(legend, line[x]));
		}
	});
}

// 2 dimensional loopage that takes world and makes it a string
World.prototype.toString = function(){
	var output = "";
	for (var y = 0; y < this.grid.height; y++){
		for (var x = 0; x <this.grid.width; X++){
			var element = this.grid.get(new Vector(x,y));
			output += charFromElement(element);
		}
		output += "\n";
	}
	return output;
};

// walls dont have any actions, they jsuttake up space
function Wall(){}

var world = new World(plan, { 	"#": Wall,
								"o": BouncingCritter});

File: content/e-life/elife.py
plan = ["####################",
		"#                  #",
		"#     #            #",
		"#            #     #",
		"#          ###     #",
		"#            #     #",
		"#     #            #",
		"#     ##           #",
		"#                  #",
		"####################"]
# in next implementation change x and y in to row/column for readings

class Vector(Object):
	def __init__(self, x, y):
		self.x = x
		self.y = y

	def add(self, other):
		return Vector(self.x + other.x, self.y + other.y)

class Grid(Object):
	def __init__(self, height, width):
		self.space = [None] * (height*width)
		self.width = width
		self.height = height

	def isInside(self, vector):
		return (0 <= vector.x <= width and 0<= vector.y <= height)

	def get(self, vector):
		return this.space[vector.x + (this.width * vector.y)]

	def set(self, vector, value):
		this.space[vector.x + (this.width * vector.y)] = value

	def forEach(self, f, context):
		for y in range(0, self.height):
			for x in range(0, self.width):
				value = self.space[x + y * self.width]
				if value is not None:
					f(context, value)

directions = {
	'n': Vector(0, -1),
	'ne': Vector(1,-1),
	'e': Vector(1, 0 ),
	'se': Vector(1,1),
	's':Vector(0,-1),
	'sw': Vector(-1,1),
	'w': vector(-1, 0),
	'nw': Vector(-1, -1)
}

def elementFromChar(legend, ch):
	if ch == "":
		return None;
	element = legend[ch]()
	element.originChar = ch
	return element

def charFromElement(element):
	if element == None:
		return ""
	else:
		return element.originChar

class World(Object):
	def __init__(self, map, legend):
		self.grid = Grid(len(map[0]), len(map))
		self.legend = legend
		# learn up on this bit..
		for y, row in enumerate(map):
			for x in range(0,len(row)):
				grid.set(Vector(x, y), elementFromChar(legend, row[x]))

	def toString(self):
		output = ""
		for y in range(0,self.grid.height):
			for x in range(0,self.grid.width):
				element = self.grid.get(Vector(x, y))
				output.append(charFromElement(element))
			output.append('\n')
		return output

	def turn (self):
		acted = [];
		self.grid.forEach(lambda (critter, vector): 	
			try:
				acted.index(critter)
			except:
				acted.append(critter)
				self.letAct(critter, vector)

	def letAct(self, critter, vector):
		action = critter.act(View(self, vector))
		if action and action.type === 'move':
			dest = self.checkDestination(action, vector)
			if dest and self.grid.get(dest) == None:
				self.grid.set(vector, None)
				self.grid.set(dest, critter)



	def checkDestination(self, action,  vector):
		if directions.hasOwnProperty(action.direction):
			dest = vector.plus(directions[action.direction])
			if self.grid.isInside(dest):
				return dest 

class View(Object):
	def __init__(self, world, vector):
		self.world = world
		self.vector = vector 

	def look(self, dir):
		target = self.vector.plus(directions[dir])
		if self.grid.isInside(target):
			return charFromElement(self.grid.isInside(target))
		else:
			return '#'

	def findAll(self, ch):
		found = []
		for dir in directions:
			if self.look(dir) is ch:
				found.append(dir)
		return found

	def find(self, ch):
		found = self.findAll(ch)
		if len(found) == 0:
			return None
		return randomElement(found)




# some critters
directionNames = "n ne e se s sw w nw"

# need to check this bidniss
def randomElement(string):
	return sample(string.split(" "))

class Wall(Object):
	def __init__(self):
		pass

class Bouncey(Object):
	def __init__(self, direction):
		self.direction = randomElement(directionNames)

	def act(self, view):
		pass
		
	





